storage-proofs/src/piece_inclusion_proof.rs

```diff
@@ -1,255 +1,473 @@
-    /// Iff it returns true, then FileInclusionProof indeed proves that piece's
-    /// bytes were included in the merkle tree corresponding to root -- and at the
-    /// position encoded in the proof.
-    fn verify(&self, root: &H::Domain, piece: &[u8]) -> bool {
-        // These checks are superfluous but inexpensive and clarifying.
-        if !(self.first_node_proof.validate::<H::Function>()
-            && self.last_node_proof.validate::<H::Function>())
-        {
-            return false;
-        }
-        // If the computed root is equal to the provided root, then the piece was provably
-        // present in the data from which the merkle tree was constructed.
-        match compute_root::<H>(&self.first_node_proof, &self.last_node_proof, piece) {
-            Ok(computed_root) => *root == computed_root,
-            Err(_) => false,
-        }
+    fn height(&self) -> usize {
+        Self::height_for_length(self.length)
     }
-}
 
-/// Compute the root which results when hashing the supplied piece_data, supplemented by the hashes
-/// in the left (first) and right (last) inclusion proofs from the FileInclusionProof.
-fn compute_root<H: Hasher>(
-    left_proof: &InclusionProof<H::Domain>,
-    right_proof: &InclusionProof<H::Domain>,
-    piece_data: &[u8],
-) -> Result<H::Domain> {
-    // Zip the left and right proof_vecs into one.
-    let proof_vecs = proof_vec(left_proof).zip(proof_vec(right_proof));
+    // `proof_length` is length of proof that comm_p is in the containing root, excluding comm_p and root, which aren't needed for the proof itself.
+    fn proof_length(&self, tree_len: usize) -> usize {
+        Self::height_for_length(tree_len) - self.height()
+    }
+
+    // Utility
+
+    pub fn piece_is_aligned(position: usize, length: usize, tree_len: usize) -> bool {
+        let capacity_at_pos = Self::subtree_capacity(position, tree_len);
 
-    let mut hasher = H::Function::default();
+        Self::is_pow2(capacity_at_pos) && capacity_at_pos >= length
+    }
 
-    let mut last_row = Vec::new();
+    fn height_for_length(n: usize) -> usize {
+        (n as f64).log2().ceil() as usize
+    }
 
-    for chunk in piece_data.chunks(32) {
-        last_row.push(H::Domain::try_from_bytes(chunk)?);
+    fn is_pow2(n: usize) -> bool {
+        n.count_ones() == 1
     }
 
-    for (height, ((l_hash, l_is_left), (r_hash, r_is_left))) in proof_vecs.enumerate() {
-        let mut row = Vec::new();
-        if !*l_is_left {
-            row.push(*l_hash);
-        };
-        row.extend(last_row);
-        if *r_is_left {
-            row.push(*r_hash);
+    fn subtree_capacity(pos: usize, total: usize) -> usize {
+        assert!(pos < total, "position must be less than tree capacity");
+
+        let mut capacity = 1;
+        // If tree is not 'full', then pos 0 will have subtree_capacity greater than size of tree.
+        let mut cursor = pos + next_pow2(total);
+
+        while cursor & 1 == 0 {
```

Comment:
Could you add a commment what the shifting and anding does here, so the intent is clearer.